proposals/ref-assignment-switch-expressions.md

```diff
@@ -0,0 +1,69 @@
+# `ref` assignment on `switch` expressions
+
+* [x] Proposed
+* [ ] Prototype: None
+* [ ] Implementation: None
+* [ ] Specification: Started, below
+
+## Summary
+[summary]: #summary
+
+Provide the ability to return `ref` values in `switch` expressions.
+
+## Motivation
+[motivation]: #motivation
+
+The `switch` expression is meant to provide a quicker way to return values based on a given expression, but does not support returning `ref` expressions. This limitation doesn't prevent any harm, and can be lifted.
+
+## Detailed design
+[design]: #detailed-design
+
+A `switch` expression may return a `ref` value. The returning expression, if not a `throw` expression, must be a `ref` expression.
+
+A `switch` expression that returns a `ref` value needs an extra `ref` in front of it when assigned/returned to a `ref` local. The examples below cover this case too. This design aligns with the current design in the ternary operator.
```

Comment:
One regret I have with conditional `ref` expressions is that the following compiles without a diagnostic: 

```csharp
cond ? ref a : ref b
```

Clearly the user meant to take a `ref` here yet they forgot the `ref` before `cond` hence ended up with a value. Guessing many devs have spent time wondering what was going wrong here. 

Would not like to repeat that mistake here. Suggest diagnostics for the following scenarios:
1. If all arms of the `switch` have `ref` but the overall expression does not.
2. If there is disagreement in the arms about `ref`

Reply:
You could(if it doesn't already exist) add an Analyzer for the `ref` expressions as well "while you're at it"(not that that's a good argument for a language decision)

Reply:
Clearly Jared has a point here, 99% of the time you want the ref of the returned expression, so a warning is very reasonable.

Adding that warning for ref ternaries should probably be done in the future, but it's off the scope of this proposal. Probably a warning wave too, to avoid breaking existing code.

Reply:
i would simply make it a requirement for switches.  WE can retcon ternaries independent of this.

Reply:
Added a simple advisory for causing a warning instead of an error that requires that the returned ref is directly used. If we settle on the error, I will update the spec accordingly.

Reply:
Update, after some thought and discussion I noticed that:
```csharp
int a = 4;
ref int b = ref a;
int c = ref a;
int d = a == b ? ref a : ref b;
```

- `c` yields error CS8171 - Cannot initialize a by-value variable with a reference
- `d` does not yield any errors and is completely legal

We should definitely strive for consistency with `ref` variables, and designing `ref` switch expressions should return an error if the value is immediately dereferenced.

As for the ref ternary expressions, it could be a breaking change, via either a warning or an error that immediately applies to existing code, hinting that it was definitely a mistake in the first place. Though discussing this one goes out of scope of this PR.

I'll prepare spec changes for this new approach on the error, would love some more input on this decision to make sure we cover this case elegantly.